### ISSUE TITLE
[C++][Qt5] Rebuild of Qt5 Code Generator

### ISSUE THREAD

**Issue by tbe** [swagger-api/swagger-codegen#5521]
The Code generated by the qt5cpp generator is incomplete and does not really match the other client library structures ( like cpprest, perl, ... ).

I would create a complete new set of templates and adapt the generator code, to make sure, the Qt5CPP generator behaves like cpprest and is more feature complete. 

Is there any active maintainer for Qt5 or such a project already in progress?

As there is also #3203 outstanding for Qt5 i would also take care of this.

**Comment by wing328** [comment]
@tbe thanks for offering help to improve the Qt5 C++ generator.

I would suggest you to base your enhancement on 2.3.0 (for breaking changes)

I've copied the contributors of Qt5 C++ generator below:

@Ragnis @fehguy @sabras75 @philicious @fortiumant @marache 

Ref: https://github.com/swagger-api/swagger-codegen/pulls?utf8=%E2%9C%93&q=is%3Apr%20label%3A%22Client%3A%20C%2B%2B%22%20is%3Aclosed%20QT5

**Comment by marache** [comment]
Hi @tbe, we have a current WIP version of a closer to Qt5 way of doing things, I'll try to propose this has a Pull Request linked to this issue. It might be a great way of starting a discussion ?

**Comment by tbe** [comment]
Hi @marache , that would be great.

@wing328, @marache  i think, using [QNetworkAuth](https://doc-snapshots.qt.io/qt5-5.8/qtnetworkauth-module.html) for handling OAuth/OAuth2 as an optional feature would be a good idea. Any thoughts about this?

We could just ignore the authentication key handling for OAuth/OAuth2 based services and allow the user either to handle the complete authentication flow with the generated code, or just supply the authkey to the underlying `QAbstractOAuth2` class.

**Comment by marache** [comment]
It's not a PR yet as we did a new plugin, I pushed the latest version here : https://github.com/cstb/qt5-swagger-codegen

**Comment by wing328** [comment]
> i think, using QNetworkAuth for handling OAuth/OAuth2 as an optional feature would be a good idea. Any thoughts about this?

Never use it myself. After reading the doc, I think we should give it a try 👍

**Comment by tbe** [comment]
I'll take the code from @marache as base, integrate the authentication handling, better namespace support ( like cpprest ) and apply the google code style where needed and possible.

**Comment by wing328** [comment]
@tbe just want to let you know that we target to release 2.3.0 by the end of July (postponed from end of May earlier)

If the refactor is ready, we'll include it in the 2.3.0 release. (this is not to say we must include the refactored Qt5 c++ client in the 2.3.0 release so please take your time with the factoring and let us know if there's anything we can help)

**Comment by wing328** [comment]
FYI. We've started publishing a SNAPSHOT version of current master (2.2.3) and 2.3.0 (a branch with breaking changes).

**Comment by antis81** [comment]
Once again I took the rocky road and went through closing all those nasty mem-leaks and thread leaks manually in the *generated* code base. Thanks to `valgrind` this is a doable task in hours instead of days… However there's not much needed to fix the default templates for good and make them "just work".

**Comment by antis81** [comment]
Do we have a `2.2.x` branch where I can request the fix to be merged? It's not completely finished/tested, code is here -> [antis81/patch-1](https://github.com/antis81/swagger-codegen/tree/patch-1)

**Comment by wing328** [comment]
@antis81 please file against 2.3.0 branch, which will be released by the end of this month.

**Comment by antis81** [comment]
@wing328 I couldn't find a 2.3.x branch to file against. Do you mean master?

**Comment by wing328** [comment]
@antis81 right, I'm referring to the master (which will be released as 2.3.0). Sorry for any confusion.

**Comment by antis81** [comment]
Alright, rebased the changes onto master and pushed. Not sure everything works as I cannot actually test it. Could someone review or even test it further? Am just about to open the PR and we see further… :)
